core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -318,9 +319,25 @@ private[spark] class SparkSubmit extends Logging {
         args.ivySettingsPath)
 
       if (!StringUtils.isBlank(resolvedMavenCoordinates)) {
-        args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
-        if (args.isPython || isInternal(args.primaryResource)) {
-          args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
+        // In K8s client mode, when in the driver, add resolved jars early as we might need
+        // them at the submit time. For example we might use the dependencies for downloading
+        // files from a Hadoop Compatible fs eg. S3. In this case the user might pass:
+        // --packages com.amazonaws:aws-java-sdk:1.7.4:org.apache.hadoop:hadoop-aws:2.7.6
+        if (isKubernetesClient &&
+          args.sparkProperties.getOrElse("spark.kubernetes.submitInDriver", "false").toBoolean) {
+          for (jar <- resolvedMavenCoordinates.split(",")) {
+            addJarToClasspath(jar, getSubmitClassLoader(sparkConf))
+          }
+        }
+        else if (isKubernetesCluster) {
```

Comment:
`isKubernetesClient` gives me enough info to know that im running in client mode.
Now given that info I can distinguish two cases a) client mode in the driver pod via cluster mode and b) pure client mode eg. running a Spark job in a pod in pure client mode.

Reply:
You reply seems to refer to my previous comment.

`isKubernetesClient` is not enough otherwise you wouldn't need to check the other config, ever.

So you have `isK8sClusterLauncher`, `isK8sClientModeDriver` or `isK8sClusterModeDriver`, basically.

Reply:
Is this comment resolved? not clear.
Move the else onto the previous line.

Reply:
Comments seems addressed, but the `else` is still in the wrong line.

Reply:
@vanzin could you please clarify why is wrong? Are you refering to this:

```
else {
          args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
          if (args.isPython || isInternal(args.primaryResource)) {
            args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
          }
```